language-plutus-core/src/Language/PlutusCore/Check/Value.hs

```diff
@@ -21,12 +22,21 @@ checkTerm
     :: (AsValueRestrictionError e tyname ann, MonadError e m) => Term tyname name ann -> m ()
 checkTerm = para $ \case
     TyAbsF ann name _ (body, rest) -> do
-        unless (isTermValue body) $
-            throwing _ValueRestrictionError $ ValueRestrictionViolation ann name
+        unless (isTermValue body) $ throwing _ValueRestrictionError $ ValueRestrictionViolation ann name
         rest
     termF                          -> traverse_ snd termF
 
 -- | Check whether a program satisfies the value restriction.
 checkProgram
     :: (AsValueRestrictionError e tyname ann, MonadError e m) => Program tyname name ann -> m ()
 checkProgram (Program _ _ term) = checkTerm term
+
+isTermValue :: Term tyname name a -> Bool
+isTermValue = isRight . termValue
+
+termValue :: Term tyname name a -> Either (NormalizationError tyname name a) (Term tyname name a)
+termValue (IWrap l pat arg term) = IWrap l pat arg <$> termValue term
+termValue (TyAbs l tn k t)       = TyAbs l tn k <$> termValue t
+termValue l@LamAbs {}            = pure l
+termValue c@Constant {}          = pure c
+termValue t                      = Left $ BadTerm (termLoc t) t "term value"
```

Comment:
Why return the same term back? Can't we do what `checkTerm` in this file does and return `m ()` for `(AsNormalizationError e tyname name ann, MonadError e m)`?

Reply:
Yeah I'm not really sure why it does this, good point.